pkg/cloud/rgraph/rnode/healthcheck/healthcheck_test.go

```diff
@@ -20,13 +20,319 @@ import (
+		t.Fatalf("hcMutRes.Freeze(_) = %v, want nil", err)
+	}
+	b := NewBuilder(id)
+	b.SetOwnership(rnode.OwnershipManaged)
+	b.SetResource(hcRes)
+	b.SetState(rnode.NodeExists)
+	node, err := b.Build()
+	if err != nil {
+		t.Fatalf("b.Build() = %v, want nil", err)
+	}
+	hcNode := node.(*healthCheckNode)
+	_, err = hcNode.resource.ToGA()
+	if err == nil {
+		t.Fatalf("hcNode.resource.ToGA() = %v, want error", err)
+	}
+	alpha, err := hcNode.resource.ToAlpha()
+	if err != nil {
+		t.Fatalf("hcNode.resource.ToAlpha() = %v, want nil", err)
+	}
+	hc := newDefaultAlphaHC()
+	if diff := cmp.Diff(alpha, &hc); diff != "" {
+		t.Fatalf("cmp.Diff(_, _) = %v, want nil", diff)
+	}
+}
+
+func buildHCNode(t *testing.T, name string, hc compute.HealthCheck) rnode.Node {
+	id := ID(projectID, meta.GlobalKey(name))
+	hcMutRes := NewMutableHealthCheck(projectID, id.Key)
+	err := hcMutRes.Access(func(x *compute.HealthCheck) {
+		*x = hc
+	})
+	if err != nil {
+		t.Fatalf("hcMutRes.Access(_) = %v, want nil", err)
+	}
+	hcRes, err := hcMutRes.Freeze()
+	if err != nil {
+		t.Fatalf("hcMutRes.Freeze(_) = %v, want nil", err)
+	}
+	b := NewBuilderWithResource(hcRes)
+	node, err := b.Build()
+	if err != nil {
+		t.Fatalf("b.Build() = %v, want nil", err)
+	}
+	return node
+}
+
+func TestHealthCheckDiff(t *testing.T) {
+	hc := newDefaultHC()
+	wantNode := buildHCNode(t, "hc-1", hc)
+
+	// compare identical nodes, no diff
+	plan, err := wantNode.Diff(wantNode)
+	if err != nil || plan.Diff != nil {
+		t.Fatalf("wantNode.Diff(wantNode) = (%v, %v), want (diff, nil)", plan.Diff, err)
+	}
+	if plan.Operation != rnode.OpNothing {
+		t.Fatalf("plan.Operation mismatch, got: %s, want %s", plan.Operation, rnode.OpNothing)
+	}
+
+	// modify health check
+	hc.CheckIntervalSec = 100
+	gotNode := buildHCNode(t, "hc-1", hc)
+
+	plan, err = wantNode.Diff(gotNode)
+	if err != nil || plan.Diff == nil {
+		t.Fatalf("wantNode.Diff(gotNode) = (%v, %v), want (diff, nil)", plan.Diff, err)
+	}
+	if plan.Operation != rnode.OpUpdate {
+		t.Fatalf("plan.Operation mismatch, got: %s, want %s", plan.Operation, rnode.OpUpdate)
+	}
+
+	//compare alpha and ga node
+	id := ID(projectID, meta.GlobalKey("hc-1"))
+	hcMutRes := NewMutableHealthCheck(projectID, id.Key)
+	err = hcMutRes.Access(func(x *compute.HealthCheck) {
+		*x = hc
+	})
+	if err != nil {
+		t.Fatalf("hcMutRes.Access(_) = %v, want nil", err)
+	}
+	err = hcMutRes.AccessAlpha(func(x *alpha.HealthCheck) {
+		x.UdpHealthCheck = &alpha.UDPHealthCheck{Port: 60}
+	})
+	if err != nil {
+		t.Fatalf("hcMutRes.AccessAlpha(_) = %v, want nil", err)
+	}
+	hcRes, err := hcMutRes.Freeze()
+	if err != nil {
+		t.Fatalf("hcMutRes.Freeze(_) = %v, want nil", err)
+	}
+	b := NewBuilderWithResource(hcRes)
+	alphaNode, err := b.Build()
+	if err != nil {
+		t.Fatalf("b.Build() = %v, want nil", err)
+	}
+	plan, err = wantNode.Diff(alphaNode)
+	if err != nil || plan.Diff == nil {
+		t.Fatalf("wantNode.Diff(wantNode) = (%v, %v), want (diff, nil)", plan.Diff, err)
```

Comment:
nit: plan can be nil if an error occurred (will get panic instead of nice error)

Reply:
if error occurs then `err != nil` will return true and `plan.Diff == nil` won't be executed, right?

Reply:
yes, I meant L267

Reply:
oh yes, god catch!